### PR TITLE
feat: npm i -g pizza

### LINKED ISSUE #27: Feature: npm install pizza
### Type of feature

🍕 Feature

### Current behavior

Is it possible to distribute this cli on npm or would that get weird? I ask because we own the [pizza](https://www.npmjs.com/package/pizza) npm package and it would be a nice name to syndicate this work from. 

I have not explored the security concerns or issues if any, but happy to add an NPM token to the Actions secrets if we want to explore this.

### Suggested solution

Looks like you do so via the goreleaser. Examples of other doing it. 

- https://blog.xendit.engineer/how-we-repurposed-npm-to-publish-and-distribute-our-go-binaries-for-internal-cli-23981b80911b
- https://blog.bywachira.com/post/create-cli-with-golang-and-publish-on-npm

### Additional context

Just look to see if its possible. 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [ ] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by Anush008**
## Description
This PR enables the application to be installed using NPM.
```sh
npm i -g pizza
```

It uses https://github.com/open-sauced/release to publish the `npm` subdirectory to NPM with semantic versioning.

The `npm` sub-directory has install and runner scripts that fetch and run the binary.

A demo release using this configuration can be found here.
https://github.com/Anush008/xdg/releases/tag/v1.10.0

The config.
https://github.com/Anush008/xdg/blob/2bcf6c2d7121ed8f77a0c31679458052fbc8b74b/.github/workflows/release.yaml#L29-L35

The published sub-directory: https://www.npmjs.com/package/xdg-anush

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [x] 🤖 Build
- [x] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
Resolves #27.

## Mobile & Desktop Screenshots/Recordings
N/A

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [x] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: npm/install.js
Comment (jpmcb) [inline npm/install.js:6]: Looks like version is being read in from the package json. So the version of the NPM package itself correlates directly to the version of the upstream software?

File: npm/install.js
Comment (jpmcb) [inline npm/install.js:12]: What is `__dirname` in this context? I'm guessing it will be the context of the global `npm i -g` command we instruct users to execute?

File: npm/install.js
Comment (jpmcb) [inline npm/install.js:14]: When the Go CLI is compiled against windows, does it automatically add the `.exe` extension? It doesn't get that automatically when I cross compile on linux.

File: .github/workflows/release.yaml
Comment (jpmcb) [reply .github/workflows/release.yaml]: Also, nit, missing newline at the end of the file.

File: npm/package.json
Comment (jpmcb) [inline npm/package.json:11]: nit: Missing newline at end of file.

File: .github/workflows/release.yaml
Comment (jpmcb) [inline .github/workflows/release.yaml:43]: Why does this need the `open-sauced/release` job to finish first? Can't they run in parallel since one is simply publishing the package to NPM and the other is actually building the binaries?

**Review by jpmcb — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yaml
Comment (jpmcb) [reply .github/workflows/release.yaml]: > Why the rename of the binary to pizzarion-cli? 

Looks like this was reverted in a commit while i was reviewing.

Please mark PRs as draft until they are ready to be reviewed to reduce churn on PR reviews! Thanks!!

Comment (Anush008) [reply .github/workflows/release.yaml]: Oh yes. I fixed it just now. I named it pizzaroni when testing this.

Comment (Anush008) [reply .github/workflows/release.yaml]: > Why the rename of the binary to `pizzarion-cli`? That doesn't seem inline with the naming in this project

Oh yes. I noticed and fixed it just now. I named it pizzaroni when testing this.

File: .github/workflows/release.yaml
Comment (jpmcb) [reply .github/workflows/release.yaml:43]: My guess is since the release action will create GitHub tags and releases for us?

File: npm/README.md
Comment (Anush008) [reply npm/README.md:1]: This is the README that'll be visible on the NPM homepage. It has to be placed within the `npm` dir. This one doesn't have the local development and `make` instructions.

File: .github/workflows/release.yaml
Comment (Anush008) [reply .github/workflows/release.yaml:43]: https://github.com/open-sauced/pizza-cli/blob/73291d13d632b709f2583d834aefe6ad758de8d7/.github/workflows/release.yaml#L66
For the binaries to be uploaded, the release tag-name is needed which is available after the completion of the release job.

File: npm/package.json
Comment (Anush008) [reply npm/package.json:7]: So `npm` symlinks `pizza` to whatever is configured as the `bin` value in this file. We expect this to run the binary inside the `bin` folder. The binary can have an `.exe` extension as well according to the platform, which is handled by the runner script.

File: npm/install.js
Comment (Anush008) [reply npm/install.js:6]: Yes. The release system will bump the version when a release is made and the bump is committed.

File: npm/install.js
Comment (Anush008) [reply npm/install.js:14]: The released windows binaries didn't have the `.exe` extension. So the install script is handling this.
https://github.com/Anush008/pizzaroni-cli/blob/6b21cb84f88f75467ce6f270e136dfca5e462d23/npm/install.js#L13-L18
https://github.com/Anush008/pizzaroni-cli/blob/6b21cb84f88f75467ce6f270e136dfca5e462d23/npm/install.js#L65
By appending the extension whenever it is downloaded from a win platform.

File: npm/package.json
Comment (Anush008) [reply npm/package.json]: https://github.com/open-sauced/pizza-cli/pull/31/commits/e51196e96f9a1bdac08190ed570cb5b62d523836

File: npm/install.js
Comment (Anush008) [reply npm/install.js:12]: The absolute path of the directory where NPM has installed our package.

File: npm/package.json
Comment (jpmcb) [reply npm/package.json:7]: > The binary can have an .exe extension as well according to the platform, which is handled by the runner script.

Gotcha. Good to know. Might have to bust out my windows box to do some testing for this 😆 great effort on getting this working across architectures. In the future, I'm ok with letting some rough edges go for windows users so we can keep moving fast on this. And hopefully some windows open source users will emerge who can help and know the platform better 👍🏼

File: npm/install.js
Comment (jpmcb) [reply npm/install.js:6]: Interesting: looks like I need to go read up on that action! So it'll automagically bump the `package.json` version for us?

**Review by jpmcb — APPROVED** [review]
Once this lands in `main` I'll cut a `beta` branch and switch over the main branch to that one.

⚠️ this may cause some disruption to people's PRs and remote clones.

File: npm/install.js
Comment (Anush008) [reply npm/install.js:6]: Yes. Here's a demo bump with the CHANGELOG.
https://github.com/Anush008/xdg/commit/d28cfd9525f9b98cbced3d84b38b809a36e836e1

**Comment by Anush008** [comment]
The existing https://www.npmjs.com/package/pizza package already has 1.0.0 published. It'll conflict with the first release I think as we can't over-publish a version.

**Comment by Anush008** [comment]
🛑 Blocked by https://github.com/open-sauced/release/issues/20.

**Review by bdougie — APPROVED** [review]
This is great work. https://github.com/open-sauced/release/issues/20 is resolved

**Comment by jpmcb** [comment]
> It'll conflict with the first release I think as we can't over-publish a version and semantic versioning for the main branch will begin with 1.0.0

😭 ah darn. I'd really, _really_ love to avoid cutting a v1 until we have to.

is there no way to blow away what's already in the NPM registry?

**Comment by bdougie** [comment]
> is there no way to blow away what's already in the NPM registry?

To prevent a left-pad there is an unpublish policy. Because there is less than 300 download it looks possible. 

https://docs.npmjs.com/policies/unpublish#packages-published-more-than-72-hours-ago

File: npm/package.json
Comment (bdougie) [reply npm/package.json:3]: @Anush008 I unpublished this, and now the namespace is available. We might want to move forward with this ASAP.

**Comment by jpmcb** [comment]
Sure thing: this is me https://www.npmjs.com/~jpmcb

**Review by jpmcb — APPROVED** [review]

**Pull request merged**
